src/firebase/api/gcp-utils.js

```diff
@@ -1,74 +1,82 @@
     return {
-      status: '500',
-      data: null,
-      error,
+      status: 401,
+      message: `not authenticated ${err}`,
+      data: {},
     };
   });
 };
 
 
 /**
  * Subscribes to pubsub to signal stimuli generation completion
+ * @param {Function} imageUrlsHandler react hook function for imageUrls
+ * @param {Function} errorHandler react hook function for error.
  */
-export const observeStimuliCompletion = async () => {
-  gcpPubsub.topic(pubsub.SIE_RESULT).onPublish((msg) => {
-    // Decode the PubSub Message body.
-    const msgBody = msg.data ?
-      Buffer.from(message.data, 'base64').toString() : null;
-    // eslint-disable-next-line no-unused-vars
-    const participantId = msg.attributes.participant_id;
-    // eslint-disable-next-line no-unused-vars
-    const experimentId = msg.attributes.experiment_id;
+export const observeStimuliCompletion =
+  async (imageUrlsHandler, errorHandler) => {
+    gcpPubsub.topic(pubsub.SIE_RESULT).onPublish((msg) => {
+      // Decode the PubSub Message body.
+      const status = msg.data ?
+        Buffer.from(message.data, 'base64').toString() : null;
+      const userId = msg.attributes.participant_id;
+      const experimentId = msg.attributes.experiment_id;
 
-    if (msgBody === 'completed') {
-      // maybe this should be handled by a react hook?
-      // call getFileUrlsFromBucket once ready to display stimuli
-    } else {
-      // likewise: maybe should be handled by a setError hook
-    }
-  });
-};
+      if (status === 'Completed') {
+        // call getFileUrlsFromBucket once ready to display stimuli
+        getFileUrlsFromBucket(userId, experimentId).then(() => {
+          imageUrlsHandler(imageUrls);
+        });
+      } else {
+        // “face_missing”, “failed”
+        // call setError() for image processing status, display on frontend
+        errorHandler(status);
+      }
+    });
+  };
 
 /**
  * Get all file urls from bucket to display in img tags
- * @param {string} bucket storage bucket
- * @param {string} bucketPrefix prefix to destination storage directory
+ * @param {String} userId user id
+ * @param {String} experimentId experiment id
+ * @return {String[]} array of processed self image urls
  */
-export const getFileUrlsFromBucket = async (bucket, bucketPrefix) => {
+const getFileUrlsFromBucket = async (userId, experimentId) => {
+  const bucketPrefix = `${userId}-${experimentId}`;
   const fileUrls = [];
-  const bucketRef = storage.ref().child(bucket).child(bucketPrefix);
-  bucketRef.getFiles((err, files) => {
+  const bucketRef = app.storage(storageBuckets.SIE_STIMULI_IMGS).ref();
+  const stimuliImagesRef = bucketRef.child(bucketPrefix);
+
+  await stimuliImagesRef.getFiles((err, files) => {
     if (!err) {
       files.forEach((file) => {
         file.getDownloadURL().then((url) => {
           fileUrls.push(url);
-        }).catch((error) => {
-          // TODO(qhoang) error handling
+        }).catch(() => {
+          return [];
```

Comment:
Since `fileUrls` is already an empty array, there's no need for this. We can just return `fileURLs`.

Reply:
this error is actually cached when **one of** the file download URL is not able to get during the for loop, thus the fileURLs array can have some URLs in it already. 

So shall we return the incomplete array or just an empty array because maybe one of the stimuli images is unable to download?

Reply:
hmm that's interesting. Does this happen often? We should do a retry (up to 3 times probably) to make sure we get all URLs